### PR TITLE
added auto updating

### PR DESCRIPTION
**Pull request by 3xela**

### CODE REVIEW

File: validator/src/updater.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    fn check_and_update(&self) -> Result<()> {
+        // Single atomic pull operation
+        let pull = Command::new("git")
+            .args(["pull", "--ff-only", "origin", &self.git_branch])
+            .output()?;
+
+        if !pull.status.success() {
+            return Err(anyhow::anyhow!("Pull failed: {}",
```
Comment (MsRandom) [inline validator/src/updater.rs]: Implement a proper error type with `thiserror`.

File: validator/src/updater.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        let stdout = String::from_utf8_lossy(&pull.stdout);
+        if stdout.contains("Already up to date") {
+            info!("No updates available");
+            return Ok(());
+        }
+
+        info!("Update found: {}", stdout.trim());
+
+        // Check if build succeeds before restarting
+        let build = Command::new("cargo")
```
Comment (MsRandom) [inline validator/src/updater.rs]: Don't build. `cargo run` builds.

File: validator/src/main.rs [hunk 3]
```diff
@@ -53,6 +57,12 @@ async fn main() {
 
     tokio::task::spawn(api_main());
 
+    let updater = Updater::new(
```
Comment (MsRandom) [inline validator/src/main.rs:58]: This needs to be used on miners as well.

**Review by MsRandom — CHANGES_REQUESTED** [review]

File: validator/src/validator/updater.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        let stdout = String::from_utf8_lossy(&pull.stdout);
+        if stdout.contains("Already up to date") {
+            info!("No updates available");
+            return Ok(());
+        }
+
+        info!("Update found: {}", stdout.trim());
+
+        let build = Command::new("cargo")
```
Comment (MsRandom) [inline validator/src/validator/updater.rs]: You don't need to run any cargo command. What I meant is that when the validator/miner restarts, they'll use `cargo run`, which requires that they build regardless

**Review by MsRandom — CHANGES_REQUESTED** [review]

File: validator/src/validator/updater.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                    error!("Update check failed: {}", e);
+                }
+                sleep(self.check_interval);
+            }
+        })
+    }
+
+    fn check_and_update(&self) -> Result<(), UpdaterError> {
+        let pull = Command::new("git")
+            .args(["pull", "--ff-only", "origin", &self.git_branch])
```
Comment (MsRandom) [inline validator/src/validator/updater.rs]: I don't think you should specify the branch. Can instead just not include `origin` and it'd use the current one

File: validator/src/validator/updater.rs (hunk 4 above)
Comment (3xela) [reply validator/src/validator/updater.rs]: that is a very good point i genuinely forgot that LOL

File: validator/src/main.rs (hunk 3 above)
Comment (3xela) [reply validator/src/main.rs:58]: line 308-312 in miner/src/main.rs, its already there

File: validator/src/validator/updater.rs (hunk 4 above)
Comment (MsRandom) [reply validator/src/validator/updater.rs]: lmao, fair

File: validator/src/main.rs (hunk 3 above)
Comment (MsRandom) [reply validator/src/main.rs:58]: You have not committed that.

Comment (3xela) [reply validator/src/main.rs:58]: ok its pushed now

File: miner/src/updater.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        if !build.status.success() {
+            return Err(UpdaterError::BuildFailed(
+                String::from_utf8_lossy(&build.stderr).to_string()
+            ));
+        }
+
+        info!("Update downloaded and built successfully. Restarting process to apply changes...");
+        Ok(())
+    }
+}
```
Comment (MsRandom) [inline miner/src/updater.rs]: Duplicating this code makes it harder to maintain. It should be moved to `base`.

**Review by MsRandom — CHANGES_REQUESTED** [review]

**Review by MsRandom — APPROVED** [review]

**Pull request merged**